text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+
+1. [non-breaking changes](#non-breaking-chanes) where the feature will be incrementally built into AWS CDK v1 without
+   affecting any existing API
+2. [breaking changes](#breaking-changes) where code generation will be changed so that statically typed languages have a
+   better experience
+
+## Non-breaking changes
+
+First, the _jsii assembly_ schema needs to be augmented so that it can represent _type unions_ as part of a library's
+exported type system. This introduces a new type `kind`, and those will coexist with _classes_, _interfaces_ and _enums_
+within the `types` section of the assembly. The required information is straight forward:
+
+| Property           | Type                        | Required     | Description                                         |
+| ------------------ | --------------------------- | ------------ | --------------------------------------------------- |
+| `name`             | `string`                    | **Required** | The unqualified name of the union type              |
+| `fqn`              | `string`                    | **Required** | The _jsii_ fully-qualified name of the type         |
+| `kind`             | `@jsii/spec.TypeKind.UNION` | **Required** | The type kind discriminator                         |
+| `assembly`         | `string`                    | **Required** | The name of the assembly containing the union type  |
+| `types`            | `string[]`                  | **Required** | The _jsii_ fully-qualified names of candidate types |
+| `namespace`        | `string`                    |              | The namespace in which the union type is declared   |
+| `docs`             | `@jsii/spec.Docs`           |              | Any documentation attached to the union type        |
+| `locationInModule` | `@jsii/spec.SourceLocation` |              | The source location of the type alias declaration   |
+
+The `jsii` compiler will then be modified in order to locate and process exported _union type_ alias declarations, such
+as:
+
+```ts
+export type UnionType = CandidateA | CandidateB;
+```
+
+In order to ensure ability to generate stable code for _union-like_ types in statically typed languages, the possible
+_candidate types_ of an aliased _type union_ must however be restricted to prevent name collisions:
+
+- all candidates **must** have distinct _unqalified names_:
+  ```ts
+  // Invalid: both types have the unqualified name `Construct`
+  export type InvalidUnion = cdk.Construct | constructs.Construct;
+  ```
+- no candidate type can have a name starting with `ListOf` or `MapOf` followed by an upper-case letter:
+  ```ts
+  // Invalid: ListOfFoo would collide with generated union-like accessors for `Foo[]`
+  export type InvalidUnion = ListOfFoo | Bar;
+  ```
+
+In order to preserve backwards-compatbility, and to retain ability to use this new feature against the current CDK v1
+codebase, the `jsii` compiler will not **require** that _type unions_ are systematically aliased. Similarly,
```

Comment:
1. The bolding is a little weird here... maybe **not require** makes more sense?
2. I think the word "systematically" can be safely removed from this sentence without changing its meaning.